Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,347 @@
+/-
+Copyright (c) 2023 Adrian Wüthrich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adrian Wüthrich
+-/
+import Mathlib.Combinatorics.SimpleGraph.Basic
+import Mathlib.Combinatorics.SimpleGraph.DegreeSum
+import Mathlib.Combinatorics.SimpleGraph.Connectivity
+import Mathlib.Combinatorics.SimpleGraph.AdjMatrix
+import Mathlib.Combinatorics.SimpleGraph.IncMatrix
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.Matrix.Rank
+import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.Fintype.Card
+import Mathlib.Data.Real.Basic
+import Mathlib.LinearAlgebra.Basic
+import Mathlib.LinearAlgebra.Matrix.BilinearForm
+import Mathlib.LinearAlgebra.Finrank
+import Mathlib.LinearAlgebra.Matrix.PosDef
+import Mathlib.LinearAlgebra.Matrix.IsDiag
+
+/-!
+# Laplacian Matrix
+
+This module defines the Laplacian matrix of a graph, and proves some of its elementary properties.
+
+## Main definitions & Results
+
+* `SimpleGraph.degMatrix`: The degree matrix of a simple graph
+* `SimpleGraph.lapMatrix `: The Laplacian matrix of a simple graph, defined as the difference
+  between the degree matrix and the adjacency matrix.
+* `isPosSemidef_lapMatrix`: The Laplacian matrix is positive semidefinite.
+* `rank_ker_lapMatrix_eq_card_ConnectedComponent`: The number of connected components in `G` is
+  the dimension of the nullspace of its Laplacian matrix.
+
+-/
+
+
+open BigOperators Finset Matrix SimpleGraph
+
+variable {V : Type*} [Fintype V] [DecidableEq V] (G : SimpleGraph V) [DecidableRel G.Adj]
+
+/-- The diagonal matrix consisting of the degrees of the vertices in the graph. -/
+def SimpleGraph.degMatrix (R : Type*) [Ring R] : Matrix V V R := Matrix.diagonal (G.degree ·)
+
+/-- `lapMatrix G R` is the matrix `L = D - A` where `D`is the degree
+  and `A` the adjacency matrix of `G`. -/
+def SimpleGraph.lapMatrix (R : Type*) [Ring R] : Matrix V V R := G.degMatrix R - G.adjMatrix R
+
+@[simp]
+theorem transpose_lapMatrix (R : Type*) [Ring R] : (G.lapMatrix R)ᵀ = G.lapMatrix R := by
+  unfold lapMatrix degMatrix
+  rw [transpose_sub, diagonal_transpose, transpose_adjMatrix]
+
+@[simp]
```

Comment:
Does tagging this as simp benefit you anywhere? I can't think of how it would but I may be missing something!

Reply:
No, I just saw that the equivalent theorem for the adjacency matrix was tagged with simp and assumed it was good practice. Should I remove @[simp] where it is not needed? I might not use them anywhere since I mostly used simp only [...].

Reply:
Hmm indeed that sounds strange to me. Well you should think about how likely it is that the lemma will be used by simp, generally lemmas that are an equality or iff statement where the LHS is more complicated than the RHS are the obvious ones. If the statement is some property like this then it is generally rarer that simp would be the tactic people would be using on goals containing that property, they'd just use that single lemma directly